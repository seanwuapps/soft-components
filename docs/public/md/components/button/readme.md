# sc-button



<!-- Auto Generated Below -->


## Usage

### Button

```html
<!-- Default -->
<sc-button onclick="alert('yo')">Alert</sc-button>

<!-- Anchor -->
<sc-button href="/" target="_blank" rel="noreferrer" title="button"
  >Anchor
</sc-button>

<!-- Block level -->
<sc-button block>Block Button</sc-button>

<!-- Icon -->
<sc-button icon>
  <i class="la la-home"></i>
</sc-button>

<!-- Bordered -->
<sc-button bordered>
  <i class="la la-back"></i> Bordered button
</sc-button>

```



## Properties

| Property   | Attribute  | Description                                                                                                                                                                                                                                                                               | Type                              | Default     |
| ---------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------- | ----------- |
| `block`    | `block`    | The button shape.                                                                                                                                                                                                                                                                         | `boolean`                         | `false`     |
| `bordered` | `bordered` | If prop exists, button will have an engraved-styled border                                                                                                                                                                                                                                | `boolean`                         | `false`     |
| `disabled` | `disabled` | If `true`, the user cannot interact with the button.                                                                                                                                                                                                                                      | `boolean`                         | `false`     |
| `download` | `download` | This attribute instructs browsers to download a URL instead of navigating to it, so the user will be prompted to save it as a local file. If the attribute has a value, it is used as the pre-filled file name in the Save prompt (the user can still change the file name if they want). | `string`                          | `undefined` |
| `href`     | `href`     | Contains a URL or a URL fragment that the hyperlink points to. If this property is set, an anchor tag will be rendered.                                                                                                                                                                   | `string`                          | `undefined` |
| `icon`     | `icon`     | Icon only button                                                                                                                                                                                                                                                                          | `boolean`                         | `false`     |
| `rel`      | `rel`      | Specifies the relationship of the target object to the link object. The value is a space-separated list of [link types](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types).                                                                                                    | `string`                          | `undefined` |
| `target`   | `target`   | Specifies where to display the linked URL. Only applies when an `href` is provided. Special keywords: `"_blank"`, `"_self"`, `"_parent"`, `"_top"`.                                                                                                                                       | `string`                          | `undefined` |
| `type`     | `type`     | The type of the button.                                                                                                                                                                                                                                                                   | `"button" \| "reset" \| "submit"` | `"button"`  |


## Events

| Event        | Description                          | Type                |
| ------------ | ------------------------------------ | ------------------- |
| `blurEvent`  | Emitted when the button loses focus. | `CustomEvent<void>` |
| `clickEvent` | Emitted when the button is clicked.  | `CustomEvent<void>` |
| `focusEvent` | Emitted when the button has focus.   | `CustomEvent<void>` |


## Slots

| Slot | Description                                                           |
| ---- | --------------------------------------------------------------------- |
|      | Content is placed between the named slots if provided without a slot. |


## CSS Custom Properties

| Name                                 | Description                                  |
| ------------------------------------ | -------------------------------------------- |
| `--sc-button-bg-color`               | Background of the button                     |
| `--sc-button-border-radius`          | Button border radius                         |
| `--sc-button-bordered-border-radius` | Bordered-style border radius                 |
| `--sc-button-bordered-border-width`  | Bordered-style border width                  |
| `--sc-button-icon-size`              | Icon button size                             |
| `--sc-button-padding-x`              | Horizontal paddings                          |
| `--sc-button-padding-y`              | Vertical paddings                            |
| `--sc-button-text-active-color`      | Text color with active prop                  |
| `--sc-button-text-color`             | Button text color                            |
| `--sc-button-text-hover-color`       | Text color on hover                          |
| `--sc-button-width`                  | Set fixed width for button (default to auto) |


## Dependencies

### Used by

 - [sc-tab-button](../tab-button)

### Graph
```mermaid
graph TD;
  sc-tab-button --> sc-button
  style sc-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Inspired by Ionic, built with StencilJS*<br>*&#10084; from Sean Wu*
