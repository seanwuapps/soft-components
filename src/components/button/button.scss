:host {
  /* button */
  --sc-button-bg-color: var(--sc-bg-color);
  --sc-button-text-color: var(--sc-text-color);
  --sc-button-border-radius: var(--sc-border-radius);
  --sc-button-padding-y: 1em;
  --sc-button-padding-x: 2em;
  --sc-button-variant-color: red;

  * {
    box-sizing: border-box;
  }

  display: inline-block;

  a {
    text-decoration: none;
    display: inline-block;
  }

  a,
  button {
    cursor: pointer;
    border: none;
    background-color: var(--sc-button-bg-color);
    color: var(--sc-button-text-color);
    border-radius: var(--sc-button-border-radius);
    font-size: 1.2em;
    @include raised();
    position: relative;
    padding: 4px;

    .button-inner {
      transition: all 0.4s ease;
      display: inline-block;
      padding: var(--sc-button-padding-y) var(--sc-button-padding-x);
      border-radius: calc(var(--sc-button-border-radius) - 4px);
    }

    &:active,
    &:hover {
      .button-inner {
        @include engraved();
      }
    }
  }
}

:host([icon]) {
  a,
  button {
    width: 3em;
    height: 3em;
    border-radius: 50%;
    text-align: center;
    line-height: calc(3em - 8px);
    padding: 4px;

    .button-inner {
      padding: 0;
      border-radius: 50%;
      width: 100%;
      height: 100%;
    }
  }
}

:host(.block) {
  display: block;

  a,
  button {
    width: 100%;

    .button-inner {
      width: 100%;
    }
  }
}
