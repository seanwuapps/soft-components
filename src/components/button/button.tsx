import {
  Component,
  Element,
  Event,
  EventEmitter,
  Host,
  Prop,
  h
} from "@stencil/core";
/**
 * @slot - Content is placed between the named slots if provided without a slot.
 */
@Component({
  tag: "sc-button",
  styleUrl: "button.scss"
  // shadow: true // disabled for form submission
})
export class Button {
  @Element() el!: HTMLElement;

  /**
   * If `true`, the user cannot interact with the button.
   */
  @Prop({ reflectToAttr: true }) disabled = false;

  /**
   * This attribute instructs browsers to download a URL instead of navigating to
   * it, so the user will be prompted to save it as a local file. If the attribute
   * has a value, it is used as the pre-filled file name in the Save prompt
   * (the user can still change the file name if they want).
   */
  @Prop() download: string | undefined;

  /**
   * Contains a URL or a URL fragment that the hyperlink points to.
   * If this property is set, an anchor tag will be rendered.
   */
  @Prop() href: string | undefined;

  /**
   * Specifies the relationship of the target object to the link object.
   * The value is a space-separated list of [link types](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types).
   */
  @Prop() rel: string | undefined;

  /**
   * The button shape.
   */
  @Prop({ reflectToAttr: true }) block?: boolean | undefined = false;

  /**
   * Icon only button
   */
  @Prop({ reflectToAttr: true }) icon?: boolean | undefined = false;

  /**
   * If prop exists, button will have an engraved-styled border
   */
  @Prop({ reflectToAttr: true }) bordered?: boolean | undefined = false;

  /**
   * Specifies where to display the linked URL.
   * Only applies when an `href` is provided.
   * Special keywords: `"_blank"`, `"_self"`, `"_parent"`, `"_top"`.
   */
  @Prop() target: string | undefined;

  /**
   * The type of the button.
   */
  @Prop() type: "submit" | "reset" | "button" = "button";

  /**
   * Emitted when the button has focus.
   */
  @Event() focusEvent!: EventEmitter<void>;

  /**
   * Emitted when the button loses focus.
   */
  @Event() blurEvent!: EventEmitter<void>;

  /**
   * Emitted when the button is clicked.
   */
  @Event() clickEvent!: EventEmitter<void>;

  private onFocus = () => {
    this.focusEvent.emit();
  };

  private onBlur = () => {
    this.blurEvent.emit();
  };

  private onClick = () => {
    this.clickEvent.emit();
  };

  render() {
    const { type, disabled, rel, target, href, icon, block, bordered } = this;
    const TagType = href === undefined ? "button" : "a";
    const attrs =
      TagType === "button"
        ? { type }
        : {
            download: this.download,
            href,
            rel,
            target
          };

    return (
      <Host
        aria-disabled={disabled ? "true" : null}
        class={{
          icon,
          block,
          bordered
        }}
      >
        <TagType
          {...attrs}
          disabled={disabled}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onClick={this.onClick}
        >
          <span class="button-inner">
            <slot name="start"></slot>
            <slot></slot>
            <slot name="end"></slot>
          </span>
        </TagType>
      </Host>
    );
  }
}
