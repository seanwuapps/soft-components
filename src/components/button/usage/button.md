<!-- Default -->
<sc-button onclick="alert('yo')">Alert</sc-button>

<!-- Anchor -->
<sc-button href="/" target="_blank" rel="noreferrer" title="button"
  >Anchor
</sc-button>

<!-- Block level -->
<sc-button block>Block Button</sc-button>

<!-- Icon -->
<sc-button icon>
  <i class="la la-home"></i>
</sc-button>

<!-- Bordered -->
<sc-button bordered>
  <i class="la la-back"></i> Bordered button
</sc-button>