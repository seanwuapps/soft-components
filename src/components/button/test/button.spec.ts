import { newSpecPage } from "@stencil/core/testing";
import { Button } from "../button";
describe("button", () => {
  it("builds", () => {
    expect(new Button()).toBeTruthy();
  });
  it("should render default button", async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<sc-button>button</sc-button>`
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it("should render a tag when href is applied", async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<sc-button href="#" target="_blank" rel="noreferrer" download>button</sc-button>`
    });
    expect(page.root).toMatchSnapshot();
  });
});
