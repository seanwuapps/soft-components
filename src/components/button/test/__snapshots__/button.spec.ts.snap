// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button should render a tag when href is applied 1`] = `
<sc-button download="" href="#" rel="noreferrer" target="_blank">
  <mock:shadow-root>
    <a download="" href="#" rel="noreferrer" target="_blank">
      <span class="button-inner">
        <slot name="start"></slot>
        <slot></slot>
        <slot name="end"></slot>
      </span>
    </a>
  </mock:shadow-root>
  button
</sc-button>
`;

exports[`button should render default button 1`] = `
<sc-button>
  <mock:shadow-root>
    <button type="button">
      <span class="button-inner">
        <slot name="start"></slot>
        <slot></slot>
        <slot name="end"></slot>
      </span>
    </button>
  </mock:shadow-root>
  button
</sc-button>
`;
